File: ios/Sources/ILDSDesignSystem/IldsBadge.swift
import ILDSTokens
import SwiftUI

public enum IldsBadgeVariant: Sendable {
    case subtle, intense, success, error, warning, info, skeleton
}

public enum IldsBadgeSize: Sendable { case small, medium, large }

public struct IldsBadge: View {
    private let label: String
    private let variant: IldsBadgeVariant
    private let size: IldsBadgeSize

    public init(_ label: String, variant: IldsBadgeVariant = .subtle, size: IldsBadgeSize = .medium) {
        self.label = label
        self.variant = variant
        self.size = size
    }

    public var body: some View {
        let colors = IldsBadgeColors.resolve(variant)
        Text(variant == .skeleton ? "   " : label)
            .font(metrics.font)
            .foregroundStyle(colors.foreground)
            .padding(.horizontal, metrics.hPad)
            .padding(.vertical, metrics.vPad)
            .background(colors.background)
            .clipShape(Capsule())
    }

    private var metrics: (font: Font, hPad: CGFloat, vPad: CGFloat) {
        switch size {
        case .small: return (.system(size: 11, weight: .medium), ILDSTokens.sp8, 2)
        case .medium: return (.system(size: ILDSTokens.fontSize12, weight: .medium), ILDSTokens.sp8, ILDSTokens.sp4)
        case .large: return (.system(size: 13, weight: .medium), ILDSTokens.sp12, ILDSTokens.sp4)
        }
    }
}

private struct IldsBadgeColors {
    let background: Color
    let foreground: Color
    static func resolve(_ v: IldsBadgeVariant) -> IldsBadgeColors {
        switch v {
        case .subtle: return .init(background: ILDSTokens.secondaryBlue50, foreground: ILDSTokens.secondaryBlue500)
        case .intense: return .init(background: ILDSTokens.secondaryBlue500, foreground: ILDSTokens.globalWhite000)
        case .success: return .init(background: ILDSTokens.successGreen500, foreground: ILDSTokens.globalWhite000)
        case .error: return .init(background: ILDSTokens.errorRed600, foreground: ILDSTokens.globalWhite000)
        case .warning: return .init(background: ILDSTokens.warningAmber500, foreground: ILDSTokens.globalWhite000)
        case .info: return .init(background: ILDSTokens.informativeBlue500, foreground: ILDSTokens.globalWhite000)
        case .skeleton: return .init(background: ILDSTokens.neutralCoolgray100, foreground: .clear)
        }
    }
}

File: android/ilds-design-system/src/main/kotlin/com/icicilombard/ilds/components/IldsBadge.kt
package com.icicilombard.ilds.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.icicilombard.ilds.tokens.IldsTokens

enum class IldsBadgeVariant { Subtle, Intense, Success, Error, Warning, Info, Skeleton }
enum class IldsBadgeSize { Small, Medium, Large }

@Composable
fun IldsBadge(
    label: String,
    modifier: Modifier = Modifier,
    variant: IldsBadgeVariant = IldsBadgeVariant.Subtle,
    size: IldsBadgeSize = IldsBadgeSize.Medium,
) {
    val colors = badgeColors(variant)
    val (hPad, vPad, style) = badgeMetrics(size)
    Text(
        text = if (variant == IldsBadgeVariant.Skeleton) "   " else label,
        modifier = modifier
            .background(colors.first, RoundedCornerShape(999.dp))
            .padding(horizontal = hPad, vertical = vPad),
        style = style,
        color = colors.second,
    )
}

private fun badgeColors(variant: IldsBadgeVariant): Pair<Color, Color> = when (variant) {
    IldsBadgeVariant.Subtle -> IldsTokens.secondaryBlue50 to IldsTokens.secondaryBlue500
    IldsBadgeVariant.Intense -> IldsTokens.secondaryBlue500 to IldsTokens.globalWhite000
    IldsBadgeVariant.Success -> IldsTokens.successGreen500 to IldsTokens.globalWhite000
    IldsBadgeVariant.Error -> IldsTokens.errorRed600 to IldsTokens.globalWhite000
    IldsBadgeVariant.Warning -> IldsTokens.warningAmber500 to IldsTokens.globalWhite000
    IldsBadgeVariant.Info -> IldsTokens.informativeBlue500 to IldsTokens.globalWhite000
    IldsBadgeVariant.Skeleton -> IldsTokens.neutralCoolgray100 to Color.Transparent
}

private fun badgeMetrics(size: IldsBadgeSize): Triple<androidx.compose.ui.unit.Dp, androidx.compose.ui.unit.Dp, TextStyle> =
    when (size) {
        IldsBadgeSize.Small -> Triple(IldsTokens.sp8, 2.dp, TextStyle(fontSize = 11.sp, fontWeight = FontWeight.Medium))
        IldsBadgeSize.Medium -> Triple(IldsTokens.sp8, IldsTokens.sp4, TextStyle(fontSize = IldsTokens.fontSize12, fontWeight = FontWeight.Medium))
        IldsBadgeSize.Large -> Triple(IldsTokens.sp12, IldsTokens.sp4, TextStyle(fontSize = 13.sp, fontWeight = FontWeight.Medium))
    }
